modules/dnn/src/layers/topk2_layer.cpp

```diff
@@ -0,0 +1,250 @@
+public:
+    int axis;
+    bool largest;
+    bool sorted;
+    int K;
+    bool dynamicK;
+
+    TopK2LayerImpl(const LayerParams& params)
+    {
+        setParamsFrom(params);
+        axis    = params.get<int>("axis", -1);
+        largest = params.get<int>("largest", 1) == 1;
+        sorted  = params.get<int>("sorted", 1) == 1;
+        CV_CheckTrue(sorted, "TopK2: sorted == false is not supported");
+        if (params.has("k")) {
+            K = params.get<int>("k");
+            CV_CheckGT(K, 0, "TopK2: K needs to be a positive integer");
+            dynamicK = false;
+        } else {
+            dynamicK = true;
+            K = 0;
+        }
+    }
+
+    bool supportBackend(int backendId) CV_OVERRIDE
+    {
+        return backendId == DNN_BACKEND_OPENCV;
+    }
+
+    bool getMemoryShapes(const std::vector<MatShape>& inputs,
+                         const int requiredOutputs,
+                         std::vector<MatShape>& outputs,
+                         std::vector<MatShape>& internals) const CV_OVERRIDE
+    {
+        CV_Assert(!inputs.empty());
+        const MatShape& inShape = inputs[0];
+        int inputDims = static_cast<int>(inShape.size());
+        int a = normalize_axis(axis, inputDims);
+        CV_Assert(a >= 0 && a < inputDims);
+
+        int outK = dynamicK ? inShape[a] : K;
+        CV_CheckLT(outK, inShape[a] + 1, "TopK2: K is out of range");
+
+        MatShape outShape = inShape;
+        outShape[a] = outK;
+        outputs.assign(2, outShape);
+        internals.clear();
+        return false;
+    }
+
+    void getTypes(const std::vector<MatType>& inputs,
+                  const int requiredOutputs,
+                  const int requiredInternals,
+                  std::vector<MatType>& outputs,
+                  std::vector<MatType>& internals) const CV_OVERRIDE
+    {
+        outputs.resize(2);
+        outputs[0] = inputs.front();
+        outputs[1] = CV_64S;
+    }
+
+private:
+    template<class Comparator, typename T>
+    void FindTopK(const Mat &input, Mat &output_vals, Mat &output_idxs, int normalized_axis)
+    {
+        const auto inShape = shape(input);
+        size_t outer = std::accumulate(inShape.begin(), inShape.begin() + normalized_axis, 1, std::multiplies<int>());
+        size_t inner = std::accumulate(inShape.begin() + normalized_axis + 1, inShape.end(), 1, std::multiplies<int>());
+        int dimAxis = inShape[normalized_axis];
+
+        auto worker = [&](const Range &r) {
+            const T* inPtr = input.ptr<T>();
+            T* valPtr = output_vals.ptr<T>();
+            int64_t* idxPtr = output_idxs.ptr<int64_t>();
+            Comparator cmp(inPtr, inner);
+            AutoBuffer<size_t> indices(dimAxis);
+            size_t* idxBuf = indices.data();
+            for (size_t b = r.start; b < r.end; ++b) {
+                for (size_t j = 0; j < inner; ++j) {
+                    size_t offset = b * dimAxis * inner + j;
+                    cmp.addOffset(offset);
```

Comment:
this is very suspicious place. Subsequent offset values inside loop form arithmetic progression (for fixed b). You add this offset to data inside cmp.addOffset(), so data addresses form geometric progression. It's wrong, they should form arithmetic progression as well. That's why I suggest you to remove addOffset and minusOffset and pass the proper data pointer to comparator constructor, which you need to place inside the inner loop.